framework/src/userobject/ElementQualityChecker.C

```diff
@@ -0,0 +1,201 @@
+                             "from libmesh.");
+
+  params.addRequiredParam<MooseEnum>("metric_type",
+                                     ElementQualityChecker::QualityMetricType(),
+                                     "Type of quality metric to be checked");
+  params.addParam<Real>("upper_bound", "the upper bound for provided metric type");
+  params.addParam<Real>("lower_bound", "The lower bound for provided metric type");
+  params.addParam<MooseEnum>("failure_type",
+                             ElementQualityChecker::FailureMessageType(),
+                             "The way how the failure of quality metric check should respond");
+  params.set<MultiMooseEnum>("execute_on") = "initial";
+
+  return params;
+}
+
+ElementQualityChecker::ElementQualityChecker(const InputParameters & parameters)
+  : ElementUserObject(parameters),
+    _m_type(getParam<MooseEnum>("metric_type").getEnum<ElemQuality>()),
+    _has_upper_bound(isParamValid("upper_bound")),
+    _has_lower_bound(isParamValid("lower_bound")),
+    _upper_bound(_has_upper_bound ? getParam<Real>("upper_bound") : 0.0),
+    _lower_bound(_has_lower_bound ? getParam<Real>("lower_bound") : 0.0),
+    _failure_type(getParam<MooseEnum>("failure_type").getEnum<FailureType>())
+{
+  if (_has_lower_bound && _has_upper_bound && _lower_bound >= _upper_bound)
+    mooseError("Provided lower bound should be less than provided upper bound!");
+}
+
+void
+ElementQualityChecker::initialize()
+{
+  _m_values.clear();
+  _elem_ids.clear();
+  _bypassed = 0;
+}
+
+void
+ElementQualityChecker::execute()
+{
+  // obtain the available quality metric for current ElemType
+  std::vector<ElemQuality> metrics_avail = libMesh::Quality::valid(_current_elem->type());
+
+  // check whether the provided quality metric is applicable to current ElemType
+  if (!checkMetricApplicability(_m_type, metrics_avail))
+  {
+    _bypassed = 1;
+    return;
+  }
+
+  std::pair<Real, Real> default_bounds = _current_elem->qual_bounds(_m_type);
+  std::pair<Real, Real> actual_bounds;
+  if (_has_lower_bound && _has_upper_bound)
+    actual_bounds = std::make_pair(_lower_bound, _upper_bound);
+  else if (_has_lower_bound)
+  {
+    if (_lower_bound < default_bounds.second)
+      actual_bounds = std::make_pair(_lower_bound, default_bounds.second);
+    else
+      mooseError("Provided lower bound should less than the default upper bound: ",
+                 default_bounds.second);
+  }
+  else if (_has_upper_bound)
+  {
+    if (_upper_bound > default_bounds.first)
+      actual_bounds = std::make_pair(default_bounds.first, _upper_bound);
+    else
+      mooseError("Provided upper bound should larger than the default lower bound: ",
+                 default_bounds.first);
+  }
+  else
+    actual_bounds = default_bounds;
+
+  // calculate and save quality metric value for current element
+  Real mv = _current_elem->quality(_m_type);
+  _m_values.insert(mv);
```

Comment:
So there's no correlation between element IDs and these values? If you just want a sorted list of elements you should probably just insert them into a vector (or list) something where the key is not used to build the data structure (since it is a real). You can always just sort once in finalize. Not only will it perform better, but it'll also just work where Reals as keys can fail.